main_loop_tf/main.py

```diff
@@ -593,89 +591,88 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                            "Tower%d_GradientNorm_%s" % (dev_idx, var_name),
+                            "%s_GradientNorm_%s" % (device, var_name),
                             tf.global_norm([grad_values])))
 
                 summaries.append(tf.summary.scalar(
-                    "Tower%d_Global_norm/clipped_grad_norm" % dev_idx,
+                    "%s_Global_norm/clipped_grad_norm" % device,
                     tf.global_norm(list(zip(*grads))[0])))
 
                 # Save gradients for each gpu to be averaged out
                 tower_grads.append(grads)
 
             # Print regularization
-            for v in tf.get_collection(
-                    tf.GraphKeys.REGULARIZATION_LOSSES):
+            for v in tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES):
                 tf.logging.debug('Regularization losses:\n{}'.format(v))
 
-    if s_gpus == cfg.num_splits:
-        num_gpus = [cfg.num_splits]
-    else:
-        # List containing the number of gpus in the full and subset case
-        # respectively at the first and second position
-        num_gpus = [cfg.num_splits, s_gpus]
-
-    (preds, labels, softmax_preds, mask, preds_flat,
-     avg_tower_loss, grads_and_vars,
-     train_op, m_iou, per_class_iou, cm_update_op,
-     reset_cm_op) = ([None] * 2 for _ in range(12))
-    update_ops = [[], []]
-    for idx, n_gpus in enumerate(num_gpus):
-        preds[idx] = tf.concat(tower_preds[:n_gpus], axis=0),
+    # Merge the towers on CPU
+    outs = {}
+    for case, this_splits in zip(['full', 'small'],
+                                 [cfg.num_splits, small_num_splits]):
+        if this_splits is None:
+            outs[case] = None
+            continue
+        preds = tf.concat(tower_preds[:this_splits], axis=0)
         # Convert from list of tensors to tensor, and average
-        softmax_preds[idx] = tf.concat(tower_soft_preds[:n_gpus], axis=0)
+        softmax_preds = tf.concat(tower_soft_preds[:this_splits], axis=0)
 
         # Concatenate the per-gpu placeholders to get a placeholder for the
         # full list of gpus and one for the subset to be used for
         # the minibatch with less batches
-        labels[idx] = tf.concat(labels_per_gpu[:n_gpus], axis=0)
+        labels = tf.concat(labels_per_gpu[:this_splits], axis=0)
 
-        # Compute the mean IoU
-        # TODO would it be better to use less precision here?
-        mask[idx] = tf.ones_like(labels[idx])
+        # Compute the (potentially masked) mean IoU
+        mask = tf.ones_like(labels)
         if len(cfg.void_labels):
-            mask[idx] = tf.cast(tf.less_equal(labels[idx], nclasses), tf.int32)
-        preds_flat[idx] = tf.reshape(preds[idx], [-1])
+            mask = tf.cast(tf.less_equal(labels, nclasses), tf.int32)
+        preds_flat = tf.reshape(preds, [-1])
+        m_iou, per_class_iou, cm_update_op, reset_cm_op = compute_mean_iou(
+            labels, preds_flat, nclasses, mask)
```

Comment:
Is the confusion matrix Variable instantiated inside the compute_mean_iou the same? It could be that since in the case of smaller batch the for loop is performed twice the confusion matrix created are two. So the update_op and reset_op could be applied to two different cm.